python/bind_data_access.h

```diff
@@ -249,30 +224,43 @@ template <class... Ts> class as_ElementArrayViewImpl {
   }
 
 private:
+  template <class Scalar, class View>
+  static auto make_scalar(Scalar &&scalar, py::object &obj, const View &view) {
+    if constexpr (std::is_same_v<std::decay_t<Scalar>,
+                                 scipp::python::PyObject>) {
+      // Returning PyObject. This increments the reference counter of
+      // the element, so it is ok if the parent `obj` (the variable)
+      // goes out of scope.
+      return scalar.to_pybind();
+    } else if constexpr (std::is_same_v<std::decay_t<Scalar>,
+                                        core::time_point>) {
+      static const auto np_datetime64 =
+          py::module::import("numpy").attr("datetime64");
+      return np_datetime64(scalar.time_since_epoch(),
+                           to_numpy_time_string(view.unit()));
+    } else if constexpr (!std::is_reference_v<Scalar>) {
+      // Views such as slices of data arrays for binned data are
+      // returned by value and require separate handling to avoid the
+      // py::return_value_policy::reference_internal in the default case
+      // below.
+      return py::cast(scalar, py::return_value_policy::move);
+    } else {
+      // Returning reference to element in variable. Return-policy
+      // reference_internal keeps alive `obj`. Note that an attempt to
+      // pass `keep_alive` as a call policy to `def_property` failed,
+      // resulting in exception from pybind11, so we have handle it by
+      // hand here.
+      return py::cast(scalar, py::return_value_policy::reference_internal, obj);
+    }
+  }
+
   // Helper function object to get a scalar value or variance.
   template <class View> struct GetScalarVisitor {
     py::object &self; // The object we're getting the value / variance from.
     std::remove_reference_t<View> &view; // self as a view.
 
     template <class Data> auto operator()(const Data &&data) const {
-      if constexpr (std::is_same_v<std::decay_t<decltype(data[0])>,
-                                   scipp::python::PyObject>) {
-        return data[0].to_pybind();
-      } else if constexpr (is_view_v<std::decay_t<decltype(data[0])>>) {
-        auto ret = py::cast(data[0], py::return_value_policy::move);
-        pybind11::detail::keep_alive_impl(ret, self);
-        return ret;
-      } else if constexpr (std::is_same_v<std::decay_t<decltype(data[0])>,
-                                          core::time_point>) {
-        static const auto np_datetime64 =
-            py::module::import("numpy").attr("datetime64");
-        return np_datetime64(data[0].time_since_epoch(),
-                             to_numpy_time_string(view.unit()));
-      } else {
-        // Passing `obj` as parent so py::keep_alive works.
-        return py::cast(data[0], py::return_value_policy::reference_internal,
-                        self);
-      }
```

Comment:
@jl-wynen You recently touched this code, do you remember if there was a reason for having the duplication here? I unified the code and tests pass.

Reply:
No, I think the code was already like this when I first touched it. I do not see a reason for the two separate implementations.